internal/kafka/internal/services/quota/ams_quota_service.go

```diff
@@ -396,3 +398,79 @@ func (q amsQuotaService) GetSubscriptionByID(subscriptionID string) (*amsv1.Subs
 
 	return resp.Body(), true, nil
 }
+
+// Checks if an organisation has a SKU entitlement and that entitlement is still active in AMS
+func (q amsQuotaService) IsQuotaEntitlementActive(kafka *dbapi.KafkaRequest) (bool, error) {
+	kafkaBillingModel, err := q.kafkaConfig.GetBillingModelByID(kafka.InstanceType, kafka.ActualKafkaBillingModel)
+	if err != nil {
+		return false, err
+	}
+
+	amsRelatedResourceBillingModel, err := q.getAMSRelatedResourceBillingModel(kafka.SubscriptionId, kafkaBillingModel)
+	if err != nil || arrays.StringEmptyPredicate(amsRelatedResourceBillingModel) {
+		return false, err
+	}
+
+	orgID, err := q.amsClient.GetOrganisationIdFromExternalId(kafka.OrganisationId)
+	if err != nil {
+		return false, err
+	}
+
+	quotaCostRelatedResourceFilter := ocm.QuotaCostRelatedResourceFilter{
+		ResourceName: &kafkaBillingModel.AMSResource,
+		Product:      &kafkaBillingModel.AMSProduct,
+		BillingModel: &amsRelatedResourceBillingModel,
+	}
+	quotaCosts, err := q.amsClient.GetQuotaCosts(orgID, true, false, quotaCostRelatedResourceFilter)
+	if err != nil {
+		return false, err
+	}
+
+	// SKU not entitled to the organisation
+	if len(quotaCosts) == 0 {
+		glog.V(10).Infof("ams quota cost not found for Kafka instance %q in organisation %q: %+v", kafka.ID, orgID, quotaCostRelatedResourceFilter)
+		return false, nil
+	}
+
+	if len(quotaCosts) > 1 {
+		glog.Warning("more than 1 quota cost was returned with the following related resource filter %+v for organisation %q", orgID, quotaCostRelatedResourceFilter)
+	}
+
+	// result will always return one quota for the given org, ams related resource billing model, resource and product for rhosak quotas
+	quotaCost := quotaCosts[0]
+
+	// when a SKU entitlement expires in AMS, the allowed value for that quota cost is set back to 0
+	// if the allowed value is 0 and consumed is greater than this, that denotes that the SKU entitlement
+	// has expired and is no longer active.
+	if quotaCost.Allowed() == 0 && quotaCost.Consumed() > quotaCost.Allowed() {
+		glog.V(10).Infof("quota %q has expired for organisation %q", quotaCost.QuotaID(), orgID)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// Determine the AMS QuotaCost related resource billing model from the AMS Subscription reserved resource billing model
+// based on the given subscription id and kafka billing model.
+func (q amsQuotaService) getAMSRelatedResourceBillingModel(subscriptionID string, kafkaBillingModel config.KafkaBillingModel) (string, error) {
+	reservedResources, err := q.amsClient.GetReservedResourcesBySubscriptionID(subscriptionID)
+	if err != nil {
+		return "", err
+	}
+
+	for _, rr := range reservedResources {
+		if kafkaBillingModel.AMSResource == rr.ResourceName() && kafkaBillingModel.HasSupportForAMSBillingModel(string(rr.BillingModel())) {
```

Comment:
Also an additional note: this is also going to be an issue if at some moment we want to remove `marketplace` from the config file as a supported AMS billing model: For old instances where reservation is made with `marketplace` it will start failing until they don't exist anymore

Reply:
Maybe we should start reserving with the appropriate non-deprecated values and remove it from the config before we introduce all the changes?
This would require verifying that it is being billed correctly with the new values

Reply:
I think i'll add some additional checks here to accommodate for that, I'm not sure how long it would take to verify those changes. 

AMS is not going to handle the migration of the billing model when they remove it? What happens if it was a long lived standard instance that wouldn't be removed? Do we need to handle moving the billing model from `marketplace` to a cloud provider specific marketplace?

Reply:
I am not sure if they will be able to ever remove it.
We were not given that information.
What I know is that we should stop using it, and if we can avoid introducing the legacy it would be better so we wouldn't need to worry about it

Reply:
I added some additional checks here to accommodate for removing marketplace in the config for now. We can then remove it once we start removing references to the 'marketplace' billing model.